internal/internal_worker_base.go

```diff
@@ -332,61 +347,100 @@ func (bw *baseWorker) runPoller() {
 	defer bw.stopWG.Done()
 	bw.metricsHandler.Counter(metrics.PollerStartCounter).Inc(1)
 
+	ctx, cancelfn := context.WithCancel(context.Background())
+	reserveChan := make(chan *SlotPermit)
+
 	for {
+		bw.stopWG.Add(1)
+		go func() {
+			defer bw.stopWG.Done()
+			s, err := bw.slotSupplier.ReserveSlot(ctx, &bw.options.slotReservationData)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					bw.logger.Error(fmt.Sprintf("Error while trying to reserve slot: %v", err))
+				} else {
+					close(reserveChan)
```

Comment:
It's rare to ever need to close channels in Go except as a broadcast mechanism. It can lead to confusing behavior. In this case the next send attempt in this loop will panic and every receive from here on out will immediately succeed with zero value

Did you maybe mean to create the channel _inside_ the loop? Even if, would recommend sending `nil` to it instead of erroring.

Reply:
I'll just send an explicit nil. In any case this is one of the things I needed to add a test for still so will do that for sure.

Reply:
Ah, so, there was a good reason for this. The explicit send of nil will block forever since, very likely, the other side waiting on it has already exited b/c of the stop channel. In fact, nothing needs to be done at all in this branch. In the branch with the logging, sending nil works.

Reply:
Do sending in a select that is bounded by context close as well